main.py

```diff
@@ -4,85 +4,111 @@
 import torch
 from torch.utils.data import DataLoader
 from model.build_BiSeNet import BiSeNet
+from model.discriminator import FCDiscriminator 
 from dataset import dataset
 import utils
 import wandb
-from train import segmentation_train
+from train import segmentation_train, adversarial_train
 from validate import val
 
+def get_dataset(dataset_name, crop_size, task, base_path):
+    path = os.path.join(base_path, dataset_name)
+    if dataset_name == 'GTA5':
+        return dataset.Gta5(path, crop_size, task)
+    elif dataset_name == 'Cityscapes':
+        return dataset.Cityscapes(path, crop_size, task)
+
+def get_optimizer(optimizer, model, learning_rate):
+    params = model.parameters()
+    if optimizer == 'rmsprop':
+        return torch.optim.RMSprop(params, learning_rate)
+    elif optimizer == 'sgd':
+        return torch.optim.SGD(params, learning_rate, momentum=.9, weight_decay=1e-4)
+    elif optimizer == 'adam':
+        return torch.optim.Adam(params, learning_rate, betas=(.9, .99))
+
 def main():
     # basic parameters
     parser = argparse.ArgumentParser()
+    parser.add_argument('--adapt_domain', action=argparse.BooleanOptionalAction, help='Source to target adaptation, disable with --no-adapt_domain')
```

Comment:
Provando a runnare il codice su colab mi dava questo errore: "AttributeError: module 'argparse' has no attribute 'BooleanOptionalAction'". Ho visto che argparse.BooleanOptionalAction è supportato solo da Python 3.9 in su, quindi io lo metterei come parametro booleano per evitare errori e per poterlo lanciare su colab